code/datums/components/trapdoor.dm

```diff
@@ -0,0 +1,230 @@
+/datum/component/trapdoor/proc/on_link_requested(datum/source, obj/item/assembly/trapdoor/assembly)
+	if(get_dist(parent, assembly) > TRAPDOOR_LINKING_SEARCH_RANGE)
+		return
+	. = LINKED_UP
+	src.assembly = assembly
+	assembly.linked = TRUE
+	UnregisterSignal(SSdcs, COMSIG_GLOB_TRAPDOOR_LINK)
+	RegisterSignal(assembly, COMSIG_ASSEMBLY_PULSED, .proc/toggle_trapdoor)
+
+///signal called by our assembly being pulsed
+/datum/component/trapdoor/proc/toggle_trapdoor(datum/source)
+	SIGNAL_HANDLER
+	if(!IS_OPEN(parent))
+		try_opening()
+	else
+		try_closing()
+
+///signal called by turf changing
+/datum/component/trapdoor/proc/turf_changed_pre(datum/source, path, new_baseturfs, flags, post_change_callbacks)
+	SIGNAL_HANDLER
+	var/turf/open/dying_trapdoor = parent
+	if(!IS_OPEN(dying_trapdoor) && !IS_OPEN(path) && path != /turf/open/floor/plating) //not a process of the trapdoor, so this trapdoor has been destroyed
+		dying_trapdoor.visible_message("<span class='warning'>the trapdoor mechanism in [dying_trapdoor] is broken!</span>")
+		if(assembly)
+			assembly.linked = FALSE
+			assembly = null
+		return
+	post_change_callbacks += CALLBACK(assembly, /obj/item/assembly/trapdoor.proc/carry_over_trapdoor, trapdoor_turf_path, stored_decals)
+
+/**
+ * # carry_over_trapdoor
+ *
+ * applies the trapdoor to the new turf (created by the last trapdoor)
+ * apparently callbacks with arguments on invoke and the callback itself have the callback args go first. interesting!
+ * change da turf my final callback. Goodbye
+ */
+/obj/item/assembly/trapdoor/proc/carry_over_trapdoor(trapdoor_turf_path, list/stored_decals, turf/new_turf)
+	new_turf.AddComponent(/datum/component/trapdoor, starts_open = FALSE, trapdoor_turf_path, src, stored_decals)
+
+
+/**
+ * # try_opening
+ *
+ * small proc for opening the turf into openspace
+ * there are no checks for opening a trapdoor, but closed has some
+ */
+/datum/component/trapdoor/proc/try_opening()
+	var/turf/open/trapdoor_turf = parent
+	trapdoor_turf.visible_message("<span class='warning'>[trapdoor_turf] swings open!</span>")
+	trapdoor_turf.ChangeTurf(/turf/open/openspace, flags = CHANGETURF_INHERIT_AIR)
+
+/**
+ * # try_closing
+ *
+ * small proc for closing the turf back into what it should be
+ * trapdoor can be blocked by building things on the openspace turf
+ */
+/datum/component/trapdoor/proc/try_closing()
+	var/turf/open/trapdoor_turf = parent
+	var/obj/structure/lattice/blocking = locate() in trapdoor_turf.contents
+	if(blocking)
+		trapdoor_turf.visible_message("<span class='warning'>The trapdoor mechanism in [trapdoor_turf] tries to shut, but is jammed by [blocking]!</span>")
+		return
+	trapdoor_turf.visible_message("<span class='warning'>The trapdoor mechanism in [trapdoor_turf] swings shut!</span>")
+	trapdoor_turf.visible_message("<span class='warning'>[trapdoor_turf] swings open!</span>")
+	trapdoor_turf.ChangeTurf(trapdoor_turf_path, flags = CHANGETURF_INHERIT_AIR)
```

Comment:
I'm not concerned about people falling to their frozen death, I am a bit worried about what this'll do to the hop line's atmos on icebox. Have you put any thought into that?

Reply:
Not TOO much. I was thinking of sticking a heater nearby, lmao

Reply:
if it's used the hop line/main hallway's gonna lock down you know

Reply:
hmm. any ideas? This may be a good enough reason to stick a hut under the trapdoor.

Reply:
I kinda agree with you. you could do something to like, block air, maybe something like a tile version of the atmos holofan things.

or just lean all the way into it and make the button freeze the hop line like a chad

Reply:
Do you even play the game? Nobody's going to use it if it locks up that whole area with firelocks while the atmosphere is returned to normal.